integrations/jira/jira/client.py

```diff
@@ -0,0 +1,116 @@
+        b64_message = b64_bytes.decode("ascii")
+        auth_value = f"Basic {b64_message}"
+
+        self.base_headers = {"Authorization": auth_value}
+
+        self.api_url = f"{self.jira_rest_url}/api/3"
+        self.webhooks_url = f"{self.jira_rest_url}/webhooks/1.0/webhook"
+
+        self.client = httpx.AsyncClient(headers=self.base_headers)
+
+    async def create_real_time_updates_webhook(self, app_host: str):
+        # webhook_target_app_host = f"{app_host}/integration/webhook"
+        # webhook_target_app_host = f"{app_host}"
+        webhook_target_app_host = "https://smee.getport.io/WIxEj7z8VIyVbUid"
+        webhook_check_response = await self.client.get(f"{self.webhooks_url}")
+        webhook_check_response.raise_for_status()
+        webhook_check = webhook_check_response.json()
+
+        for webhook in webhook_check:
+            if webhook["url"] == webhook_target_app_host:
+                logger.info("Ocean real time reporting webhook already exists")
+                return
+
+        body = {
+            "name": WEBHOOK_NAME,
+            "url": webhook_target_app_host,
+            "events": WEBHOOK_EVENTS,
+            "filters": {"issue-related-events-section": ""},
+        }
+
+        webhook_create_response = await self.client.post(
+            f"{self.webhooks_url}", json=body
+        )
+        webhook_create_response.raise_for_status()
+        logger.info("Ocean real time reporting webhook created")
+
+    async def get_single_project(self, project_key):
+        project_response = await self.client.get(
+            f"{self.api_url}/project/{project_key}"
+        )
+        project_response.raise_for_status()
+        return project_response.json()
+
+    async def get_all_projects(self):
+        project_response = await self.client.get(f"{self.api_url}/project")
+        project_response.raise_for_status()
+
+        return project_response.json()
+
+    async def get_single_issue(self, issue_key):
+        issue_response = await self.client.get(f"{self.api_url}/issue/{issue_key}")
+        issue_response.raise_for_status()
+        return issue_response.json()
+
+    async def get_paginated_issues(self):
+        logger.info("Getting issues from Jira")
+
+        params = {
+            "maxResults": 0,
+            "startAt": 0,
+        }
+
+        config = typing.cast(JiraResourceConfig, event.resource_config)
+
+        if config.selector.jql:
+            params["jql"] = config.selector.jql
+
+        total_issues = (await self._get_paginated_issues(params))["total"]
+
+        params["maxResults"] = PAGE_SIZE
+        while params["startAt"] <= total_issues:
+            logger.info(f"Current query position: {params['startAt']}/{total_issues}")
+            issue_response_list = (await self._get_paginated_issues(params))["issues"]
+            yield issue_response_list
+            params["startAt"] += PAGE_SIZE
+
+    async def _get_paginated_issues(self, params):
```

Comment:
Public functions should be at the bottom

Reply:
Moved